File: PRM-VAULTPOOL-feature-authentication-java/android/app/src/main/java/com/lavelahotel/poolbooking/ui/auth/FirebaseAuthViewModel.java
package com.lavelahotel.poolbooking.ui.auth;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseUser;
import com.lavelahotel.poolbooking.data.repository.FirebaseAuthRepository;

public class FirebaseAuthViewModel extends ViewModel {

    private final FirebaseAuthRepository repository;
    private final MutableLiveData<AuthUiState> uiState =
            new MutableLiveData<>(AuthUiState.Idle.INSTANCE);

    public FirebaseAuthViewModel(FirebaseAuthRepository repository) {
        this.repository = repository;
    }

    public LiveData<AuthUiState> getUiState() {
        return uiState;
    }

    public void register(String email, String password, String fullName) {
        uiState.setValue(AuthUiState.Loading.INSTANCE);

        repository.registerWithEmail(email, password, fullName, new FirebaseAuthRepository.AuthCallback() {
            @Override
            public void onSuccess(FirebaseUser user) {
                uiState.postValue(new AuthUiState.RegisterSuccess(
                        "Đăng ký thành công. Vui lòng kiểm tra email để xác thực.",
                        user
                ));
            }

            @Override
            public void onError(Throwable t) {
                uiState.postValue(new AuthUiState.Error(getErrorMessage(t)));
            }
        });
    }

    public void login(String email, String password) {
        uiState.setValue(AuthUiState.Loading.INSTANCE);

        repository.loginWithEmail(email, password, new FirebaseAuthRepository.AuthCallback() {
            @Override
            public void onSuccess(FirebaseUser user) {
                repository.reloadUser(new FirebaseAuthRepository.AuthCallback() {
                    @Override
                    public void onSuccess(FirebaseUser reloadedUser) {
                        if (reloadedUser.isEmailVerified()) {
                            uiState.postValue(new AuthUiState.LoginSuccess(reloadedUser));
                        } else {
                            uiState.postValue(new AuthUiState.EmailNotVerified(
                                    "Vui lòng xác thực email trước khi đăng nhập.",
                                    reloadedUser
                            ));
                        }
                    }

                    @Override
                    public void onError(Throwable t) {
                        // Nếu reload thất bại nhưng login thành công, vẫn cho login
                        uiState.postValue(new AuthUiState.LoginSuccess(user));
                    }
                });
            }

            @Override
            public void onError(Throwable t) {
                uiState.postValue(new AuthUiState.Error(getErrorMessage(t)));
            }
        });
    }

    public void resendVerificationEmail() {
        uiState.setValue(AuthUiState.Loading.INSTANCE);
        repository.resendVerificationEmail(new FirebaseAuthRepository.SimpleCallback() {
            @Override
            public void onSuccess() {
                uiState.postValue(new AuthUiState.EmailSent(
                        "Email xác thực đã được gửi. Vui lòng kiểm tra hộp thư."
                ));
            }

            @Override
            public void onError(Throwable t) {
                uiState.postValue(new AuthUiState.Error(getErrorMessage(t)));
            }
        });
    }

    public void checkEmailVerification() {
        repository.reloadUser(new FirebaseAuthRepository.AuthCallback() {
            @Override
            public void onSuccess(FirebaseUser user) {
                if (user.isEmailVerified()) {
                    uiState.postValue(new AuthUiState.EmailVerified(user));
                } else {
                    uiState.postValue(new AuthUiState.EmailNotVerified(
                            "Email chưa được xác thực.",
                            user
                    ));
                }
            }

            @Override
            public void onError(Throwable t) {
                uiState.postValue(new AuthUiState.Error(getErrorMessage(t)));
            }
        });
    }

    public void resetState() {
        uiState.setValue(AuthUiState.Idle.INSTANCE);
    }

    private String getErrorMessage(Throwable throwable) {
        String message = throwable != null ? throwable.getMessage() : null;
        if (message == null) {
            return "Đã xảy ra lỗi. Vui lòng thử lại sau.";
        }

        switch (message) {
            case "The email address is already in use by another account.":
                return "Email này đã được sử dụng. Vui lòng sử dụng email khác.";
            case "The password is too weak.":
                return "Mật khẩu quá yếu. Vui lòng sử dụng mật khẩu mạnh hơn.";
            case "The email address is badly formatted.":
                return "Email không hợp lệ.";
            case "There is no user record corresponding to this identifier.":
                return "Email hoặc mật khẩu không đúng.";
            case "The password is invalid or the user does not have a password.":
                return "Mật khẩu không đúng.";
            case "A network error (such as timeout, interrupted connection or unreachable host) has occurred.":
                return "Lỗi kết nối. Vui lòng kiểm tra internet và thử lại.";
            default:
                return message != null ? message : "Đã xảy ra lỗi. Vui lòng thử lại sau.";
        }
    }
}

File: android/app/src/main/java/com/lavelahotel/poolbooking/ui/auth/FirebaseAuthViewModel.kt
package com.lavelahotel.poolbooking.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseUser
import com.lavelahotel.poolbooking.data.repository.FirebaseAuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class FirebaseAuthViewModel(
    private val repository: FirebaseAuthRepository
) : ViewModel() {
    
    private val _uiState = MutableStateFlow<AuthUiState>(AuthUiState.Idle)
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()
    
    fun register(email: String, password: String, fullName: String) {
        viewModelScope.launch {
            _uiState.value = AuthUiState.Loading
            repository.registerWithEmail(email, password, fullName)
                .onSuccess { user ->
                    _uiState.value = AuthUiState.RegisterSuccess(
                        message = "Đăng ký thành công. Vui lòng kiểm tra email để xác thực.",
                        user = user
                    )
                }
                .onFailure {
                    _uiState.value = AuthUiState.Error(getErrorMessage(it))
                }
        }
    }
    
    fun login(email: String, password: String) {
        viewModelScope.launch {
            _uiState.value = AuthUiState.Loading
            repository.loginWithEmail(email, password)
                .onSuccess { user ->
                    // Reload user để check emailVerified status
                    repository.reloadUser()
                        .onSuccess { reloadedUser ->
                            if (reloadedUser.isEmailVerified) {
                                _uiState.value = AuthUiState.LoginSuccess(reloadedUser)
                            } else {
                                _uiState.value = AuthUiState.EmailNotVerified(
                                    message = "Vui lòng xác thực email trước khi đăng nhập.",
                                    user = reloadedUser
                                )
                            }
                        }
                        .onFailure {
                            _uiState.value = AuthUiState.LoginSuccess(user)
                        }
                }
                .onFailure {
                    _uiState.value = AuthUiState.Error(getErrorMessage(it))
                }
        }
    }
    
    fun resendVerificationEmail() {
        viewModelScope.launch {
            _uiState.value = AuthUiState.Loading
            repository.resendVerificationEmail()
                .onSuccess {
                    _uiState.value = AuthUiState.EmailSent(
                        message = "Email xác thực đã được gửi. Vui lòng kiểm tra hộp thư."
                    )
                }
                .onFailure {
                    _uiState.value = AuthUiState.Error(getErrorMessage(it))
                }
        }
    }
    
    fun checkEmailVerification() {
        viewModelScope.launch {
            repository.reloadUser()
                .onSuccess { user ->
                    if (user.isEmailVerified) {
                        _uiState.value = AuthUiState.EmailVerified(user)
                    } else {
                        _uiState.value = AuthUiState.EmailNotVerified(
                            message = "Email chưa được xác thực.",
                            user = user
                        )
                    }
                }
                .onFailure {
                    _uiState.value = AuthUiState.Error(getErrorMessage(it))
                }
        }
    }
    
    fun resetState() {
        _uiState.value = AuthUiState.Idle
    }
    
    private fun getErrorMessage(throwable: Throwable): String {
        return when (throwable.message) {
            "The email address is already in use by another account." -> 
                "Email này đã được sử dụng. Vui lòng sử dụng email khác."
            "The password is too weak." -> 
                "Mật khẩu quá yếu. Vui lòng sử dụng mật khẩu mạnh hơn."
            "The email address is badly formatted." -> 
                "Email không hợp lệ."
            "There is no user record corresponding to this identifier." -> 
                "Email hoặc mật khẩu không đúng."
            "The password is invalid or the user does not have a password." -> 
                "Mật khẩu không đúng."
            "A network error (such as timeout, interrupted connection or unreachable host) has occurred." -> 
                "Lỗi kết nối. Vui lòng kiểm tra internet và thử lại."
            else -> throwable.message ?: "Đã xảy ra lỗi. Vui lòng thử lại sau."
        }
    }
}

sealed class AuthUiState {
    object Idle : AuthUiState()
    object Loading : AuthUiState()
    data class RegisterSuccess(val message: String, val user: FirebaseUser) : AuthUiState()
    data class LoginSuccess(val user: FirebaseUser) : AuthUiState()
    data class EmailNotVerified(val message: String, val user: FirebaseUser) : AuthUiState()
    data class EmailVerified(val user: FirebaseUser) : AuthUiState()
    data class EmailSent(val message: String) : AuthUiState()
    data class Error(val message: String) : AuthUiState()
}
